File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortableDecoderPropertyTests.cs
using System.Buffers.Binary;
using System.Collections.Immutable;
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>
/// PB6: the decoders hold their contract on arbitrary and mutated input, inside deterministic bounds.
/// </summary>
/// <remarks>
/// Every vector so far presents input a person wrote. These present input nobody wrote, which is the
/// only way to check the property the vectors assume everywhere: that a decoder either produces a
/// value or refuses with a portable category, and never does anything else. A foreign exception
/// escaping here would cross the seam as a runtime type in exactly the way C4 forbids.
///
/// The generator is seeded and the iteration count fixed, so a failure is reproducible and the suite
/// is not a source of intermittent red. That is what "within deterministic bounds" asks for: this is
/// a property test, not a soak test.
/// </remarks>
public sealed class PortableDecoderPropertyTests
{
    private const int Seed = 0x_50_17_1D_E1;
    private const int Iterations = 2000;

    /// <summary>
    /// The only two outcomes a decoder may produce. Anything else is the defect being hunted.
    /// </summary>
    private static void RequireDecoderContract(Action decode, string what)
    {
        try
        {
            decode();
        }
        catch (PortableFaultException)
        {
            // A refusal carrying a portable category is the contract.
        }
        catch (PortableProcessFailureException)
        {
            // A locally observed loss is the other permitted outcome.
        }
        catch (Exception exception)
        {
            Assert.Fail(
                $"{what} escaped as {exception.GetType().FullName} rather than a portable refusal: {exception.Message}");
        }
    }

    [Test]
    public void Arbitrary_bytes_are_either_decoded_or_refused_with_a_portable_category()
    {
        var random = new Random(Seed);
        var buffer = new byte[64];

        for (var iteration = 0; iteration < Iterations; iteration++)
        {
            random.NextBytes(buffer);
            var length = random.Next(0, buffer.Length);
            var input = buffer.AsSpan(0, length).ToArray();

            RequireDecoderContract(
                () => PortableCbor.Decode(input, PortableLimits.Declared),
                $"Decoding {length} arbitrary bytes");
        }
    }

    [Test]
    public void Arbitrary_bytes_are_never_accepted_as_a_well_formed_envelope()
    {
        var random = new Random(Seed + 1);
        var buffer = new byte[96];
        var accepted = 0;

        for (var iteration = 0; iteration < Iterations; iteration++)
        {
            random.NextBytes(buffer);
            var input = buffer.AsSpan(0, random.Next(0, buffer.Length)).ToArray();

            try
            {
                PortableEnvelopeCodec.Decode(input, PortableLimits.Declared);
                accepted++;
            }
            catch (PortableFaultException)
            {
            }
            catch (PortableProcessFailureException)
            {
            }
            catch (Exception exception)
            {
                Assert.Fail($"An envelope decode escaped as {exception.GetType().FullName}: {exception.Message}");
            }
        }

        // Random bytes are not a contract-versioned envelope with a declared kind and a channel
        // identity. Accepting one would mean the envelope's required fields are not required.
        Assert.That(accepted, Is.Zero, "Random input was accepted as a well-formed envelope.");
    }

    /// <summary>
    /// Single-byte mutations of a valid frame. This reaches the decode paths random bytes never do,
    /// because a mutant is still nearly a legal message.
    /// </summary>
    [Test]
    public void Every_single_byte_mutation_of_a_valid_frame_is_refused_or_decoded_but_never_escapes()
    {
        var valid = PortableEnvelopeCodec.Encode(PortableEnvelope.Control(
            PortableEnvelopeKind.Ready,
            PortableChannelId.New()));
        var random = new Random(Seed + 2);

        for (var index = 0; index < valid.Length; index++)
        {
            foreach (var replacement in new byte[] { 0x00, 0xFF, 0x7B, (byte)random.Next(256) })
            {
                var mutant = valid.ToArray();
                mutant[index] = replacement;
                RequireDecoderContract(
                    () => PortableEnvelopeCodec.Decode(mutant, PortableLimits.Declared),
                    $"Byte {index} replaced with 0x{replacement:x2}");
            }
        }
    }

    /// <summary>Truncations of a valid frame, which are the interrupted-frame shape at every offset.</summary>
    [Test]
    public void Every_truncation_of_a_valid_frame_is_refused_rather_than_half_decoded()
    {
        var valid = PortableEnvelopeCodec.Encode(PortableEnvelope.Control(
            PortableEnvelopeKind.Ready,
            PortableChannelId.New()));

        for (var length = 0; length < valid.Length; length++)
        {
            var truncated = valid.AsSpan(0, length).ToArray();
            RequireDecoderContract(
                () => PortableEnvelopeCodec.Decode(truncated, PortableLimits.Declared),
                $"A frame truncated to {length} bytes");
        }
    }

    /// <summary>
    /// Nesting is bounded before it becomes recursion. A generator that can always add one more
    /// level is the honest test of a declared depth limit.
    /// </summary>
    [Test]
    public void Nesting_beyond_the_declared_depth_is_refused_at_every_depth_past_the_bound()
    {
        foreach (var depth in new[]
        {
            PortableLimits.Declared.MaxNestingDepth + 1,
            PortableLimits.Declared.MaxNestingDepth + 8,
            PortableLimits.Declared.MaxNestingDepth * 4,
            10_000
        })
        {
            var nested = new byte[depth + 1];
            Array.Fill(nested, (byte)0x81, 0, depth);
            nested[^1] = 0xF6;

            var fault = Assert.Throws<PortableFaultException>(
                () => PortableCbor.Decode(nested, PortableLimits.Declared),
                $"Depth {depth} did not refuse.");
            Assert.That(fault!.Category, Is.EqualTo(PortableProtocolCategory.LimitExceeded));
        }
    }

    /// <summary>
    /// A length prefix is refused on the prefix alone, so a hostile declaration never causes an
    /// allocation proportional to it.
    /// </summary>
    [Test]
    public void A_hostile_length_prefix_is_refused_before_any_allocation()
    {
        var random = new Random(Seed + 3);

        for (var iteration = 0; iteration < 256; iteration++)
        {
            var declared = (uint)random.NextInt64(PortableLimits.Declared.MaxFrameBytes + 1, uint.MaxValue);
            var prefix = new byte[4];
            BinaryPrimitives.WriteUInt32BigEndian(prefix, declared);
            using var stream = new MemoryStream(prefix);

            var fault = Assert.ThrowsAsync<PortableFaultException>(async () =>
                await PortableFraming.ReadFrameAsync(stream, PortableLimits.Declared, CancellationToken.None));

            Assert.Multiple(() =>
            {
                Assert.That(fault!.Category, Is.EqualTo(PortableProtocolCategory.LimitExceeded));
                Assert.That(stream.Position, Is.EqualTo(4), "The declared body was never read.");
            });
        }
    }

    /// <summary>
    /// A refusal never carries a runtime type, a stack trace, or a namespace, whatever the input.
    /// </summary>
    /// <remarks>
    /// This is the property CH-18 states, checked against input nobody chose rather than against the
    /// handful of messages a person wrote.
    /// </remarks>
    [Test]
    public void No_refusal_message_carries_a_runtime_type_or_a_stack_trace()
    {
        var random = new Random(Seed + 4);
        var buffer = new byte[80];
        var forbidden = new[] { "Brontide.", "System.", "Exception", "   at ", "`1", "+<" };
        var observed = ImmutableArray.CreateBuilder<string>();

        for (var iteration = 0; iteration < Iterations; iteration++)
        {
            random.NextBytes(buffer);
            var input = buffer.AsSpan(0, random.Next(0, buffer.Length)).ToArray();

            try
            {
                PortableEnvelopeCodec.Decode(input, PortableLimits.Declared);
            }
            catch (PortableFaultException fault)
            {
                observed.Add($"{fault.LocalCode}|{fault.Message}");
            }
            catch (PortableProcessFailureException failure)
            {
                observed.Add(failure.Message);
            }
        }

        Assert.Multiple(() =>
        {
            Assert.That(observed, Is.Not.Empty, "The generator produced no refusals to inspect.");
            foreach (var message in observed.Distinct())
            {
                foreach (var marker in forbidden)
                {
                    Assert.That(
                        message,
                        Does.Not.Contain(marker),
                        $"A refusal leaked '{marker}': {message}");
                }
            }
        });
    }
}

File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortableDecoderPropertyTests.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open System
open System.IO
open System.Threading
open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// PB6: the decoders hold their contract on arbitrary and mutated input, inside deterministic bounds.
///
/// Every vector so far presents input a person wrote. These present input nobody wrote, which is the
/// only way to check the property the vectors assume everywhere.
///
/// This stack states the property more strongly than the Reference one can. Refusals here are
/// returned values rather than raised failures, so the claim is not "it raises only the right
/// exception" but "it does not raise at all": every input, however malformed, produces an `Ok` or an
/// `Error` carrying a portable category. A raised exception is a defect by construction, not merely
/// the wrong kind of refusal.
///
/// The generator is seeded and the iteration count fixed, so a failure is reproducible and the suite
/// is not a source of intermittent red. That is what "within deterministic bounds" asks for: this is
/// a property test, not a soak test.
[<TestFixture>]
type PortableDecoderPropertyTests() =

    let seed = 0x50171DE1
    let iterations = 2000

    /// Decoding returns a result. Anything raised is the defect being hunted.
    let decodeWithoutRaising what (decode: unit -> PortableResult<'T>) =
        try
            decode ()
        with exn ->
            Assert.Fail $"{what} raised {exn.GetType().FullName} rather than returning a refusal: {exn.Message}"
            failwith "unreachable"

    let arbitrary (random: Random) (size: int) =
        let buffer = Array.zeroCreate<byte> size
        random.NextBytes buffer
        buffer[.. random.Next(0, size) - 1]

    [<Test>]
    member _.``arbitrary bytes are either decoded or refused, and never raise``() =
        let random = Random seed

        for _ in 1..iterations do
            let input = arbitrary random 64

            decodeWithoutRaising "Decoding arbitrary bytes" (fun () -> PortableCbor.decode input PortableLimits.declared)
            |> ignore

    [<Test>]
    member _.``arbitrary bytes are never accepted as a well-formed envelope``() =
        let random = Random(seed + 1)
        let mutable accepted = 0

        for _ in 1..iterations do
            let input = arbitrary random 96

            match decodeWithoutRaising "An envelope decode" (fun () -> EnvelopeCodec.decode input PortableLimits.declared) with
            | Ok _ -> accepted <- accepted + 1
            | Error _ -> ()

        // Random bytes are not a contract-versioned envelope with a declared kind and a channel
        // identity. Accepting one would mean the envelope's required fields are not required.
        Assert.That(accepted, Is.Zero, "Random input was accepted as a well-formed envelope.")

    /// Single-byte mutations of a valid frame. This reaches the decode paths random bytes never do,
    /// because a mutant is still nearly a legal message.
    [<Test>]
    member _.``every single-byte mutation of a valid frame is refused or decoded but never raises``() =
        let valid =
            expectOk (EnvelopeCodec.encode (Envelope.control EnvelopeKind.Ready (ChannelId.next ()) (CborMap [])))

        let random = Random(seed + 2)

        for index in 0 .. valid.Length - 1 do
            for replacement in [ 0uy; 0xFFuy; 0x7Buy; byte (random.Next 256) ] do
                let mutant = Array.copy valid
                mutant[index] <- replacement

                decodeWithoutRaising
                    $"Byte {index} replaced with 0x%02x{replacement}"
                    (fun () -> EnvelopeCodec.decode mutant PortableLimits.declared)
                |> ignore

    /// Truncations of a valid frame, which are the interrupted-frame shape at every offset.
    [<Test>]
    member _.``every truncation of a valid frame is refused rather than half-decoded``() =
        let valid =
            expectOk (EnvelopeCodec.encode (Envelope.control EnvelopeKind.Ready (ChannelId.next ()) (CborMap [])))

        for length in 0 .. valid.Length - 1 do
            decodeWithoutRaising
                $"A frame truncated to {length} bytes"
                (fun () -> EnvelopeCodec.decode valid[.. length - 1] PortableLimits.declared)
            |> ignore

    /// Nesting is bounded before it becomes recursion. A generator that can always add one more
    /// level is the honest test of a declared depth limit.
    [<Test>]
    member _.``nesting beyond the declared depth is refused at every depth past the bound``() =
        for depth in
            [ PortableLimits.declared.MaxNestingDepth + 1
              PortableLimits.declared.MaxNestingDepth + 8
              PortableLimits.declared.MaxNestingDepth * 4
              10_000 ] do
            let nested = Array.create (depth + 1) 0x81uy
            nested[depth] <- 0xF6uy

            let fault =
                decodeWithoutRaising $"Depth {depth}" (fun () -> PortableCbor.decode nested PortableLimits.declared)
                |> expectRefusal

            Assert.That(fault.Category, Is.EqualTo ProtocolCategory.LimitExceeded, $"Depth {depth} did not refuse.")

    /// A length prefix is refused on the prefix alone, so a hostile declaration never causes an
    /// allocation proportional to it.
    [<Test>]
    member _.``a hostile length prefix is refused before any allocation``() =
        let random = Random(seed + 3)

        for _ in 1..256 do
            let declared =
                uint32 (random.NextInt64(int64 PortableLimits.declared.MaxFrameBytes + 1L, int64 UInt32.MaxValue))

            let prefix = Array.zeroCreate<byte> 4
            System.Buffers.Binary.BinaryPrimitives.WriteUInt32BigEndian(Span prefix, declared)
            use stream = new MemoryStream(prefix)

            let fault =
                (PortableFraming.readFrame stream PortableLimits.declared CancellationToken.None).Result
                |> expectRefusal

            assertAll (fun () ->
                Assert.That(fault.Category, Is.EqualTo ProtocolCategory.LimitExceeded)
                // The declared body was never read.
                Assert.That(stream.Position, Is.EqualTo 4L))

    /// A refusal never carries a runtime type, a stack trace, or a namespace, whatever the input.
    ///
    /// This is the property CH-18 states, checked against input nobody chose rather than against the
    /// handful of messages a person wrote.
    [<Test>]
    member _.``no refusal message carries a runtime type or a stack trace``() =
        let random = Random(seed + 4)
        let forbidden = [ "Brontide."; "Microsoft.FSharp"; "System."; "Exception"; "   at " ]
        let observed = ResizeArray<string>()

        for _ in 1..iterations do
            let input = arbitrary random 80

            match decodeWithoutRaising "An envelope decode" (fun () -> EnvelopeCodec.decode input PortableLimits.declared) with
            | Error(Refused fault) -> observed.Add $"{fault.LocalCode}|{fault.Message}"
            | Error(Interrupted failure) -> observed.Add failure.Message
            | Ok _ -> ()

        assertAll (fun () ->
            Assert.That(observed, Is.Not.Empty, "The generator produced no refusals to inspect.")

            for message in Seq.distinct observed do
                for marker in forbidden do
                    Assert.That(message, Does.Not.Contain marker, $"A refusal leaked '{marker}': {message}"))
